streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java

```diff
@@ -314,7 +348,7 @@ void initializeStoreOffsetsFromCheckpoint(final boolean storeDirIsEmpty) {
     }
 
     private void maybeRegisterStoreWithChangelogReader(final String storeName) {
-        if (isLoggingEnabled(storeName)) {
+        if (isLoggingEnabled(storeName) && changelogReader != null) {
```

Comment:
This sort of thing makes me a bit nervous. At the very least, I think we should throw an exception if any of these methods are invoked on a ProcessorStateManager for an unassigned task with uninitialized fields.

To be honest the safest thing to do would probably be to split up the ProcessorStateManager and extract everything from before/after assignment into separate classes (or subclasses). I won't demand this if it's going to be a huge refactor, just give it a thought

Reply:
> This sort of thing makes me a bit nervous. At the very least, I think we should throw an exception if any of these methods are invoked on a ProcessorStateManager for an unassigned task with uninitialized fields.

Reply:
> This sort of thing makes me a bit nervous. At the very least, I think we should throw an exception if any of these methods are invoked on a ProcessorStateManager for an unassigned task with uninitialized fields.

This is a tricky one. This method is called from `registerStore` and `registerStateStores`, which both need to be called for our "startup tasks" to initialize the stores.

If we wanted to make calling this method from startup tasks an error, we'd need to move this conditional out into both `registerStore` and `registerStateStores`, or find an alternative way to conditionally call this method.

One simple solution would be to create an "no-op" implementation of `ChangelogReader` that we use in our startup tasks, before replacing with a "real" one once they've been assigned. Would this be a better approach?

Reply:
WRT refactoring ProcessorStateManager: that might be a good idea in the longer term, although I'd like to avoid doing it here, to limit the size and scope of this PR.